### PR TITLE
Add annotations and labels to the Spec.

### LINKED ISSUE #108: Labels and extension meta data in containers
I would like to discuss ability to attach application-specific meta data to container images. Here are a couple of examples that come to mind:

* bookeeping of versions of libraries included in the image that can be later inspected by security audit tools to update containers with obsolete/vulnerable libraries
* adding application-specific information that can be used by implementation to execute custom actions, e.g. host restrictions (this container can only run on a host with > 60 GB or RAM)
* adding custom signatures to the app-specific content in the image

For example, app container spec adresses this use-case by introducing labels:

> labels (list of objects, optional) used during image discovery and dependency resolution. The listed objects must have two key-value pairs: name is restricted to the AC Identifier formatting and value is an arbitrary string. Label names must be unique within the list, and (to avoid confusion with the image's name) cannot be "name". Several well-known labels are defined:
    version when combined with "name", this SHOULD be unique for every build of an app (on a given "os"/"arch" combination).
    os, arch can together be considered to describe the syscall ABI this image requires. arch is meaningful only if os is provided. If one or both values are not provided, the image is assumed to be OS- and/or architecture-independent. Currently supported combinations are listed in the types.ValidOSArch variable, which can be updated by an implementation that supports other combinations. The combinations whitelisted by default are (in format os/arch): linux/amd64, linux/i386, freebsd/amd64, freebsd/i386, freebsd/arm, darwin/x86_64, darwin/i386. See the Operating System spec for the environment apps can expect to run in given a known os label.

After trying to use labels in practice, I can say that ACI's labels have the following limitations:

* Value of the label is limited to string, what makes it hard to include structured data in the value
* Label names are unique, but not enforced in the data structure that is presented as a list.

The useful part about labels is namespaces, that helps to identify the purpose of meta data and avoid collisions.

It would be very helpful to include the mechanism of adding meta data to the open container images that addresses some of the limitations listed above.

### PR DESCRIPTION
**Pull request by vishh**
Fixes #108

### CODE REVIEW

**Comment by mrunalp** [comment]
Do we need Labels and Annotations in both spec and state? I thought Labels were more of a runtime grouping mechanism so belong to the state and Annotations static and hence should be in the spec config.

**Comment by vbatts** [comment]
@mrunalp the labels are useful in both places. And as far as annotations, there may be runtimes that need to stash runtime specific metadata, and it would provide a clean place to do so.

**Comment by wking** [comment]
It looks like the runtime will be ignoring ‘annotations’ [1].  Why not
follow the approach #188 proposes and use a single, opaque field
(‘interface{}’ in Go and ‘JSON’ or ‘object’ in the Markdown [2])?
Then the higher-level tools that will be consuming this information
can specify their own schema, Go types, etc.  I don't see a benefit to
requiring a particular schema for data that the runtime will ignore,
and requring a map[string]string seems like it conflicts with “This
information may be … structured” [1].

I'm -1 on labels (which do impact the runtime interface) until we have
clarity on that interface.  For example, the ‘list’ command used as an
example [3] is not one of our current operations [4].  If we want to
add label-based identifiers for stop and exec, I'd rather have links
from those actions to the labels documentation.  And “This operation
MUST generate an error if it is not provided the container ID” [5]
would probably need to be softened.

[1]: https://github.com/vishh/specs/commit/ce688a750f5de242a2d0572e860876b0e42612d7#diff-c9c91c29b41257aea3a3403cc606ad99R298
[2]: https://github.com/opencontainers/specs/pull/188/files#r45020143
[3]: https://github.com/vishh/specs/commit/ce688a750f5de242a2d0572e860876b0e42612d7#diff-c9c91c29b41257aea3a3403cc606ad99R293
[4]: https://github.com/opencontainers/specs/blob/a1e32a8ead2ba57adce3e36e956b4dc32c1b85c4/runtime.md#operations
[5]: https://github.com/opencontainers/specs/blob/a1e32a8ead2ba57adce3e36e956b4dc32c1b85c4/runtime.md#stop

**Comment by vishh** [comment]
How much do we care about the `state` from a Spec perspective?

**Comment by wking** [comment]
On Fri, Mar 04, 2016 at 01:54:50PM -0800, Vish Kannan wrote:
> How much do we care about the `state` from a Spec perspective?

I care because it is the key that allows container access without
going through an established runtime operation [1].  If you need to
adjust something that doesn't have an OCI knob, attach some monitoring
that doesn't have an OCI integration, etc., the state gives you the
platform primitives you need to interact with the container directly.

[1]: https://github.com/opencontainers/specs/blob/a1e32a8ead2ba57adce3e36e956b4dc32c1b85c4/runtime.md#operations

**Comment by vishh** [comment]
@philips @crosbymichael @mrunalp @vbatts: How do we want to move forward on this PR?

**Comment by mrunalp** [comment]
@vishh Could you give use cases for each as we had discussed in the call? I think that will help move this.

**Comment by crosbymichael** [comment]
I'm still unclear why we need to.  its like 2x the complexity and ppl could be confused when and why to use each one.

**Comment by wking** [comment]
On Tue, Mar 08, 2016 at 10:42:38AM -0800, Michael Crosby wrote:
> I'm still unclear why we need to.  its like 2x the complexity and
> ppl could be confused when and why to use each one.

One is for opaque-to-the-runtime data, and the other is for grouping
(as defined in this PR).  I think that's a clear enough distinction.
I'd prefer folks used other files for opaque-to-the-runtime data, but
I'm ok with a blessed field in config.json for opaque-to-the-runtime
data [1].

[1]: https://groups.google.com/a/opencontainers.org/d/msg/dev/p_u2xps6jvo/bcMI6hm1DwAJ
     Subject: Re: Labels and extension meta data in containers #108
     Date: Wed, 30 Dec 2015 20:54:37 -0800
     Message-ID: <20151231045437.GA6362@odin.tremily.us>

**Comment by vishh** [comment]
Updates PR to not include Labels. Removed Annotations from `state.go`

**Comment by crosbymichael** [comment]
LGTM

**Comment by mrunalp** [comment]
LGTM

File: config.go [hunk 1]
```diff
@@ -18,6 +18,8 @@ type Spec struct {
 	Mounts []Mount `json:"mounts"`
 	// Hooks are the commands run at various lifecycle events of the container.
 	Hooks Hooks `json:"hooks"`
+	// Annotations is an unstructured key value map that may be set by external tools to store and retrieve arbitrary metadata.
+	Annotations map[string]string `json:"annotations,omitempty"`
```
Comment (wking) [inline config.go]: Are we sure we don't [want an opaque type here][1]?  Or will we just revisit that later when somebody needs something that doesn't fit into a `map[string]string`?

[1]: https://github.com/opencontainers/specs/pull/331#issuecomment-192477931

File: config.md [hunk 2]
```diff
@@ -231,4 +231,17 @@ If a hook returns a non-zero exit code, then an error is logged and the remainin
 `args` and `env` are optional.
 The semantics are the same as `Path`, `Args` and `Env` in [golang Cmd](https://golang.org/pkg/os/exec/#Cmd).
 
+## Annotations
+
+Annotations are optional arbitrary non-identifying metadata that can be attached to containers.
```
Comment (wking) [inline config.md]: “arbitrary” and ”non-identifying” seem to conflict.  I'd expect we want to drop “non-identifying” and just missed it in the reroll that dropped labels.

**Pull request merged**